src/library_openal.js

```diff
@@ -1,1622 +1,4591 @@
+            // Sample the current time as late as possible to mitigate drift
+            startTime = Math.max(startTime, src.context.audioCtx.currentTime);
+            audioSrc.start(startTime, startOffset);
+          } else if (typeof(audioSrc.noteOn) !== "undefined") {
+            startTime = Math.max(startTime, src.context.audioCtx.currentTime);
+            audioSrc.noteOn(startTime);
+#if OPENAL_DEBUG
+            if (offset > 0.0) {
+              Runtime.warnOnce("The current browser does not support AudioBufferSourceNode.start(when, offset); method, so cannot play back audio with an offset "+startOffset+" secs! Audio glitches will occur!");
             }
 #endif
           }
 #if OPENAL_DEBUG
           else {
-            Runtime.warnOnce('Unable to start AudioBufferSourceNode playback! Not supported by the browser?');
+            Runtime.warnOnce("Unable to start AudioBufferSourceNode playback! Not supported by the browser?");
           }
 
-          console.log('updateSource queuing buffer ' + i + ' for source ' + idx + ' at ' + startTime + ' (offset by ' + offset + ')');
+          console.log("scheduleSourceAudio() queuing buffer " + buf.id + " for source " + src.id + " at " + startTime + " (offset by " + startOffset + ")");
 #endif
+          audioSrc._startTime = startTime;
+          src.audioQueue.push(audioSrc);
+
+          startTime += duration;
         }
 
-        startTime = endTime;
+        startOffset = 0.0;
+        bufCursor++;
       }
     },
 
-    setSourceState: function setSourceState(src, state) {
-#if OPENAL_DEBUG
-      var idx = AL.srcIdBySrc(src);
-#endif
+    // Advance the state of a source forward to the current time
+    updateSourceTime: function(src) {
+      var currentTime = src.context.audioCtx.currentTime;
+      if (src.state !== 0x1012 /* AL_PLAYING */) {
+        return currentTime;
+      }
+
+      // if the start time is unset, determine it based on the current offset
+      if (!isFinite(src.bufStartTime)) {
+        src.bufStartTime = currentTime - src.bufOffset / src.playbackRate;
+        src.bufOffset = 0.0;
+      }
+
+      var nextStartTime = 0.0;
+      while (src.audioQueue.length) {
+        var audioSrc = src.audioQueue[0];
+        src.bufsProcessed += audioSrc._skipCount;
+        nextStartTime = audioSrc._startTime + audioSrc._duration; // n.b. audioSrc._duration already factors in playbackRate, so no divide by src.playbackRate on it.
+
+        if (currentTime < nextStartTime) {
+          break;
+        }
+
+        src.audioQueue.shift();
+        src.bufStartTime = nextStartTime;
+        src.bufOffset = 0.0;
+        src.bufsProcessed++;
+      }
+
+      var audioSrc = src.audioQueue[0];
+      if (audioSrc) {
+        src.bufOffset = (currentTime - audioSrc._startTime) * src.playbackRate;
+      } else {
+        var skipCount = 0;
+        // In the absence of scheduled buffers, fast-forward playback state to the current time
+        while (true) {
```

Comment:
This `while(true)` is making me nervous because after staring at the loop body for 15 minutes, I cannot easily statically conclude that the loop will terminate in all possible cases. Could this be a `for(var i = 0; i < src.bufQueue.length; ++i)` loop instead? If I understand correctly, we only need to iterate as many times as there are queued buffers for a source?